### PR TITLE
Extend discord messages with content

### PR DESCRIPTION
**Pull request by mogoll92**
These changes are aimed to support discord content parameter besides with embedded message and allow to overcome limitation when you need to ping someone or group of people. (Ping over embedded messages doesn't trigger discord notifications)

Alertmanager already has struct that allows to send content.
```
type webhook struct {
	Content string         `json:"content"`
	Embeds  []webhookEmbed `json:"embeds"`
}
```

These changes make content optional and allow to send any text that doesn't break discord limits with embedded message.
To enable this user should apply the following changes:
```
receivers:
  - name: nothing
  - name: discord-success
    discord_configs:
      - webhook_url: ""
        content: '{{ template "discord.content" . }}'
        title: '{{ template "discord.title" . }}'
        message: '{{ template "discord.message" . }}'
 ```
 
 and override text in custom template like the following.
 ```
 {{ define "discord.content" }}
Hey you! <@6532081833221312329>
{{ end }}
 ```
 
 Here is the result.
Screenshot 2024-08-30 at 15 34 41

### CODE REVIEW

**Comment by grobinson-grafana** [comment]
Hello! 👋 Thank you for opening a PR with this change. However, I think it's a duplicate of #3821?

**Comment by grobinson-grafana** [comment]
Hello! 👋 I think this might be a duplicate of https://github.com/prometheus/alertmanager/pull/3821?

**Comment by mogoll92** [comment]
> Hello! 👋 I think this might be a duplicate of #3821?

Yeah, I saw that PR but looks like author are not interested in pushing it forward.

**Comment by grobinson-grafana** [comment]
Sounds good to me! Before I review, it looks like you have somehow brought in a number of additional commits that aren't relevant to your PR. For example, commits `fbcb49b` and `c07eec8`. Can you fix your PR to remove those additional commits? I think just we need commits `ae07f25` and `5875042`.

**Comment by grobinson-grafana** [comment]
Hi, looks like you added even more commits?

**Comment by mogoll92** [comment]
> Hi, looks like you added even more commits?

yeah, a bit struggled on replacing with needed set of commits. I will ping you here once I'm done.

**Comment by mogoll92** [comment]
@grobinson-grafana check if nothing needs to add, please.

**Comment by featheredtoast** [comment]
Is this still being looked at? I am interested in the avatar/username fields, and it'd be nice for this work to be in place before chipping away at adding other fields.

File: template/default.tmpl [hunk 1]
```diff
@@ -123,6 +123,7 @@ Alerts Resolved:
 {{ end }}
 {{ end }}
 
+{{ define "discord.default.content" }}{{ end }}
```
Comment (grobinson-grafana) [inline template/default.tmpl:126]: You want to update `DefaultDiscordConfig` on Line 46 of `config/notifiers.go` so this template is used.

**Comment by grobinson-grafana** [comment]
I've tested it and looks good to me! I've left a comment of something small that's missing. Please also add some documentation for this in `docs/configuration.md`.

Something that I'm still not 100% satisfied with is the intuitiveness of `title`, `message` and `content`. It seems like users will be confused when to use `title` and `message` and when to use `content`.

Since the use case here is mentioning users, and nothing else, it might make more sense to do the following instead:

```yml
receivers:
  - name: discord
    discord_configs:
      - webhook_url: https://discord.com/api/webhooks/abc
        title: '{{ template "discord.title" . }}'
        message: '{{ template "discord.message" . }}'
        mentions:
          - 6532081833221312329
```

This will avoid the confusion mentioned and be more intuitive for users.

**Comment by grobinson-grafana** [comment]
I've tested it and looks good to me! I've left a comment of something small that's missing. Please also add some documentation for this in `docs/configuration.md`.

Something that I'm still not 100% satisfied with is the intuitiveness of `title`, `message` and `content`. It seems like users will be confused when to use `title` and `message` and when to use `content`.

Since the use case here is mentioning users, and nothing else, it might make more sense to do the following instead:

```yml
receivers:
  - name: discord
    discord_configs:
      - webhook_url: https://discord.com/api/webhooks/abc
        title: '{{ template "discord.title" . }}'
        message: '{{ template "discord.message" . }}'
        mentions:
          - 6532081833221312329
```

This will avoid the confusion and be more intuitive to users while still solving the intended use case.

**Comment by mogoll92** [comment]
> I've tested it and looks good to me! I've left a comment of something small that's missing. Please also add some documentation for this in `docs/configuration.md`.
> 
> Something that I'm still not 100% satisfied with is the intuitiveness of `title`, `message` and `content`. It seems like users will be confused when to use `title` and `message` and when to use `content`.
> 
> Since the use case here is mentioning users, and nothing else, it might make more sense to do the following instead:
> 
> ```yaml
> receivers:
>   - name: discord
>     discord_configs:
>       - webhook_url: https://discord.com/api/webhooks/abc
>         title: '{{ template "discord.title" . }}'
>         message: '{{ template "discord.message" . }}'
>         mentions:
>           - 6532081833221312329
> ```
> 
> This will avoid the confusion and be more intuitive to users while still solving the intended use case.

It will confuse even more, as there is no mentions entity in discord, mentions is some kind of feature that you. can use in context, but it isn't the same. Overall in discord the embedded message it's a part of the message and can't exist separately. If you would like to have mirror between alertmanager discord configuration and discord's entities I would recommend the following structure.

```
receivers:
  - name: discord
    discord_configs:
      - webhook_url: https://discord.com/api/webhooks/abc
        context: '{{ template "discord.content" . }}'
        embedded:
            title: '{{ template "discord.title" . }}'
            message: '{{ template "discord.message" . }}'
```
In this case it represent exactly what it is. Also this structure could be extended with other fields that discord supports.

**Review by gotjosh — APPROVED** [review]
LGTM

I'll add the documentation in a follow up PR.

**Comment by gotjosh** [comment]
Hey @mogoll92  I'll give you about ~30 minutes or so to see if you can get it sorted and update the documentation. Otherwise, I'll proceed with #4080 unless you feel oppose to it.

**Comment by mogoll92** [comment]
> Hey @mogoll92 I'll give you about ~30 minutes or so to see if you can get it sorted and update the documentation. Otherwise, I'll proceed with #4080 unless you feel oppose to it.

if it's okay to you feel free to proceed with it, no objections at all.

**Review by grobinson-grafana — APPROVED** [review]

**Comment by gotjosh** [comment]
Can you rebase and update the vfs_asserts?

File: docs/configuration.md [hunk 2]
```diff
@@ -883,6 +883,9 @@ webhook_url_file: <filepath>
 # Message body template.
 [ message: <tmpl_string> | default = '{{ template "discord.default.message" . }}' ]
 
+# Message content template. Limited with 2000 characters.
```
Comment (grobinson-grafana) [inline docs/configuration.md:886]:
```suggestion
# Message content template. Limited to 2000 characters.
```

I think is better grammar. But we can fix this afterwards.

**Comment by mogoll92** [comment]
> Thank you very much for your contribution!

you are welcome, first of all I need those changes so I'm interested in it. Thought it never will be happened since so much time passed with no response :)

**Pull request merged**